### ISSUE TITLE
f2py needs a main program

### ISSUE THREAD

**Issue by pochedls** [numpy/numpy#14222]
This is a clarification about the behavior of f2py. In some environments it seems like a main program must be defined in the Fortran program, but in the documentation and in some environments (e.g., example [here](https://numpy.org/devdocs/f2py/usage.html)) it is possible to only define subroutines.  

Should I need to create a (sometimes arbitrary) main program in my Fortran code or is this a bug?

### Reproducing code example:

Fortran code: 
```fortran
SUBROUTINE foo; 
PRINT *,'Hello world'; 
END SUBROUTINE
```

```
f2py -c -m test test.f90
```

### Error message:

> running build
> running config_cc
> unifing config_cc, config, build_clib, build_ext, build commands --compiler options
> running config_fc
> unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
> running build_src
> build_src
> building extension "test" sources
> f2py options: []
> f2py:> /tmp/tmpi3w3wjme/src.linux-x86_64-3.7/testmodule.c
> creating /tmp/tmpi3w3wjme/src.linux-x86_64-3.7
> Reading fortran codes...
> 	Reading file 'test.f90' (format:free)
> Post-processing...
> 	Block: test
> 			Block: foo
> Post-processing (stage 2)...
> Building modules...
> 	Building module "test"...
> 		Constructing wrapper function "foo"...
> 		  foo()
> 	Wrote C/API module "test" to file "/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/testmodule.c"
>   adding '/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.c' to sources.
>   adding '/tmp/tmpi3w3wjme/src.linux-x86_64-3.7' to include_dirs.
> copying /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.c -> /tmp/tmpi3w3wjme/src.linux-x86_64-3.7
> copying /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.h -> /tmp/tmpi3w3wjme/src.linux-x86_64-3.7
> build_src: building npy-pkg config files
> running build_ext
> customize UnixCCompiler
> C compiler: /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC
> 
> creating /tmp/tmpfz7g67_x/tmp
> creating /tmp/tmpfz7g67_x/tmp/tmpfz7g67_x
> compile options: '-MMD -MF /tmp/tmpfz7g67_x/file.c.d -c'
> x86_64-conda_cos6-linux-gnu-cc: /tmp/tmpfz7g67_x/file.c
> customize UnixCCompiler using build_ext
> get_default_fcompiler: matching types: '['gnu95', 'intel', 'lahey', 'pg', 'absoft', 'nag', 'vast', 'compaq', 'intele', 'intelem', 'gnu', 'g95', 'pathf95', 'nagfor']'
> customize Gnu95FCompiler
> Found executable /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-gfortran
> Found executable /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-ld
> Found executable /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-ar
> Found executable /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-ranlib
> /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/distutils/fcompiler/__init__.py:517: UserWarning: FFLAGS is used as is, not appended to flags already defined by numpy.distutils! Use NPY_DISTUTILS_APPEND_FLAGS=1 to obtain appending behavior instead (this behavior will become default in a future release).
>   fflags = self.flag_vars.flags + dflags + oflags + aflags
> /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/distutils/fcompiler/__init__.py:530: UserWarning: LDFLAGS is used as is, not appended to flags already defined by numpy.distutils! Use NPY_DISTUTILS_APPEND_FLAGS=1 to obtain appending behavior instead (this behavior will become default in a future release).
>   linker_so_flags = self.flag_vars.linker_so
> /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/distutils/fcompiler/__init__.py:540: UserWarning: LDFLAGS is used as is, not appended to flags already defined by numpy.distutils! Use NPY_DISTUTILS_APPEND_FLAGS=1 to obtain appending behavior instead (this behavior will become default in a future release).
>   linker_exe_flags = self.flag_vars.linker_exe
> customize Gnu95FCompiler
> customize Gnu95FCompiler using build_ext
> building 'test' extension
> compiling C sources
> C compiler: /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC
> 
> creating /tmp/tmpi3w3wjme/tmp
> creating /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme
> creating /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme/src.linux-x86_64-3.7
> compile options: '-I/tmp/tmpi3w3wjme/src.linux-x86_64-3.7 -I/home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include -I/home/pc/bin/anaconda2/envs/rttov123/include/python3.7m -c'
> x86_64-conda_cos6-linux-gnu-cc: /tmp/tmpi3w3wjme/src.linux-x86_64-3.7/testmodule.c
> x86_64-conda_cos6-linux-gnu-cc: /tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.c
> In file included from /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1830:0,
>                  from /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
>                  from /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
>                  from /tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.h:13,
>                  from /tmp/tmpi3w3wjme/src.linux-x86_64-3.7/testmodule.c:15:
> /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
>  #warning "Using deprecated NumPy API, disable it with " \
>   ^~~~~~~
> In file included from /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1830:0,
>                  from /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
>                  from /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
>                  from /tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.h:13,
>                  from /tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.c:2:
> /home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
>  #warning "Using deprecated NumPy API, disable it with " \
>   ^~~~~~~
> compiling Fortran sources
> Fortran f77 compiler: /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-gfortran -Wall -g -ffixed-form -fno-second-underscore -fopenmp -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -O3 -funroll-loops
> Fortran f90 compiler: /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-gfortran -Wall -g -fno-second-underscore -fopenmp -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -O3 -funroll-loops
> Fortran fix compiler: /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-gfortran -Wall -g -ffixed-form -fno-second-underscore -Wall -g -fno-second-underscore -fopenmp -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -O3 -funroll-loops
> compile options: '-I/tmp/tmpi3w3wjme/src.linux-x86_64-3.7 -I/home/pc/bin/anaconda2/envs/rttov123/lib/python3.7/site-packages/numpy/core/include -I/home/pc/bin/anaconda2/envs/rttov123/include/python3.7m -c'
> x86_64-conda_cos6-linux-gnu-gfortran:f90: test.f90
> /home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-gfortran -Wall -g -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/testmodule.o /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.o /tmp/tmpi3w3wjme/test.o -L/home/pc/bin/anaconda2/envs/rttov123/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/../lib -L/home/pc/bin/anaconda2/envs/rttov123/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/../lib -lgfortran -o ./test.cpython-37m-x86_64-linux-gnu.so
> /home/pc/bin/anaconda2/envs/rttov123/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pc/bin/anaconda2/envs/rttov123/bin/../x86_64-conda_cos6-linux-gnu/sysroot/usr/lib/../lib/Scrt1.o: in function `_start':
> /home/rdonnelly/mc/conda-bld/compilers_linux-64_1534865402226/work/.build/src/glibc-2.12.2/csu/../sysdeps/x86_64/elf/start.S:99: undefined reference to `main'
> collect2: error: ld returned 1 exit status
> error: Command "/home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-gfortran -Wall -g -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/testmodule.o /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.o /tmp/tmpi3w3wjme/test.o -L/home/pc/bin/anaconda2/envs/rttov123/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/../lib -L/home/pc/bin/anaconda2/envs/rttov123/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/../lib -lgfortran -o ./test.cpython-37m-x86_64-linux-gnu.so" failed with exit status 1
> 

### Other notes:

I can get f2py to work with this Fortran code: 

```fortran
PROGRAM MAIN
PRINT *,'Hello world'
END PROGRAM MAIN

SUBROUTINE foo;
PRINT *,'Hello world'
END SUBROUTINE
```

Though in this particular environment, I get `ImportError: dynamic module does not define module export function (PyInit_hello)` on import. Using a more basic environment (e.g., `conda create -n test -c conda-forge python=3 numpy`), the code compiles and can be imported. 

### Numpy/Python version information:

1.17.0 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)
[GCC 7.3.0]

[Created with: `conda create -n rttov123 -c conda-forge python=3 hdf5 gfortran_linux-64 netcdf4 netcdf-fortran`]

**Comment by durack1** [comment]
@pochedls thanks for submitting this, I am wondering about this now too

**Comment by charris** [comment]
@pearu Thoughts?

**Comment by charris** [comment]
What environments are causing problems?

**Comment by pv** [comment]
Maybe: `LDFLAGS` environment variable set, but does not contain `-shared`.

**Comment by durack1** [comment]
@charris @pv we're trying to understand this, if you create a vanilla conda env with just numpy, what compiler do you get? Just the system fortran, or does numpy bring a compiler along for the env ride?

**Comment by pochedls** [comment]
Thanks for looking at this. I think this is the minimal environment to reproduce the issue (issue arises when adding gfortran_linux-64 to my conda environment).

|Environment|numpy version|f2py result (no main program)|f2py result (with main)|python test|
|-|-|-|-|-|
|`conda create -n test -c conda-forge python=3 numpy`|1.17.0|Produces .so file|Produces .so file|Module works|
|Last row + `gfortran_linux-64`|1.17.0|Similar error as outlined in main issue|Produces .so file|ImportError: dynamic module does not define module export function (PyInit_hello)|

@pv - I do not have anything set for `LD_FLAGS`

@durack1 `echo $FC` yields: `/export_backup/pochedley1/bin/anaconda2/envs/test/bin/x86_64-conda_cos6-linux-gnu-gfortran`

**Comment by durack1** [comment]
@pochedls what is the `Fortran compilers found` output, so mine (RHEL6.10) for a recently generated env is:
```
(amipbcs190730) duro@ocean:[~]:[12987]> python
Python 2.7.15 | packaged by conda-forge | (default, Jul  2 2019, 00:39:44) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy.distutils.fcompiler as fcompiler
>>> fcompiler.show_fcompilers()
Warning: Can't read registry to find the necessary compiler setting
Make sure that Python modules _winreg, win32api or win32con are installed.
Gnu95FCompiler instance properties:
  archiver        = ['/usr/bin/gfortran', '-cr']
  compile_switch  = '-c'
  compiler_f77    = ['/usr/bin/gfortran', '-Wall', '-g', '-ffixed-form', '-
                    fno-second-underscore', '-fPIC', '-O3', '-funroll-loops']
  compiler_f90    = ['/usr/bin/gfortran', '-Wall', '-g', '-fno-second-
                    underscore', '-fPIC', '-O3', '-funroll-loops']
  compiler_fix    = ['/usr/bin/gfortran', '-Wall', '-g', '-ffixed-form', '-
                    fno-second-underscore', '-Wall', '-g', '-fno-second-
                    underscore', '-fPIC', '-O3', '-funroll-loops']
  libraries       = ['gfortran']
  library_dirs    = ['/usr/lib/gcc/x86_64-redhat-linux/4.4.7',
                    '/usr/lib/gcc/x86_64-redhat-linux/4.4.7']
  linker_exe      = ['/usr/bin/gfortran', '-Wall', '-Wall']
  linker_so       = ['/usr/bin/gfortran', '-Wall', '-g', '-Wall', '-g', '-
                    shared']
  object_switch   = '-o '
  ranlib          = ['/usr/bin/gfortran']
  version         = LooseVersion ('4.4.7')
  version_cmd     = ['/usr/bin/gfortran', '-dumpversion']
Fortran compilers found:
  --fcompiler=gnu95  GNU Fortran 95 compiler (4.4.7)
Compilers available for this platform, but not found:
  --fcompiler=absoft   Absoft Corp Fortran Compiler
  --fcompiler=compaq   Compaq Fortran Compiler
  --fcompiler=g95      G95 Fortran Compiler
  --fcompiler=gnu      GNU Fortran 77 compiler
  --fcompiler=intel    Intel Fortran Compiler for 32-bit apps
  --fcompiler=intele   Intel Fortran Compiler for Itanium apps
  --fcompiler=intelem  Intel Fortran Compiler for 64-bit apps
  --fcompiler=lahey    Lahey/Fujitsu Fortran 95 Compiler
  --fcompiler=nag      NAGWare Fortran 95 Compiler
  --fcompiler=nagfor   NAG Fortran Compiler
  --fcompiler=pathf95  PathScale Fortran Compiler
  --fcompiler=pg       Portland Group Fortran Compiler
  --fcompiler=vast     Pacific-Sierra Research Fortran 90 Compiler
Compilers not available on this platform:
  --fcompiler=flang     Portland Group Fortran LLVM Compiler
  --fcompiler=hpux      HP Fortran 90 Compiler
  --fcompiler=ibm       IBM XL Fortran Compiler
  --fcompiler=intelev   Intel Visual Fortran Compiler for Itanium apps
  --fcompiler=intelv    Intel Visual Fortran Compiler for 32-bit apps
  --fcompiler=intelvem  Intel Visual Fortran Compiler for 64-bit apps
  --fcompiler=mips      MIPSpro Fortran Compiler
  --fcompiler=none      Fake Fortran compiler
  --fcompiler=sun       Sun or Forte Fortran 95 Compiler
For compiler details, run 'config_fc --verbose' setup command.
>>> import numpy
>>> print(numpy.version.version)
1.16.4
```

**Comment by pearu** [comment]
> Should I need to create a (sometimes arbitrary) main program in my Fortran code or is this a bug?

No, main program is not required in your Fortran code as extension modules (that f2py also generates) are shared libraries.

As a fix, make sure that `-shared` option is used when linking the extension module.

In your case, `-shared` is not used. Try running (if you have the build files available)
```
/home/pc/bin/anaconda2/envs/rttov123/bin/x86_64-conda_cos6-linux-gnu-gfortran -shared -Wall -g -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/testmodule.o /tmp/tmpi3w3wjme/tmp/tmpi3w3wjme/src.linux-x86_64-3.7/fortranobject.o /tmp/tmpi3w3wjme/test.o -L/home/pc/bin/anaconda2/envs/rttov123/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/../lib -L/home/pc/bin/anaconda2/envs/rttov123/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/../lib -lgfortran -o ./test.cpython-37m-x86_64-linux-gnu.so
```
and then try importing the `test` module. Notice that I just copied the linker command from your output and inserted the `-shared` option.

HTH

PS: This is not a f2py bug.

**Comment by pochedls** [comment]
Thanks @pearu - this solved my problem. In my minimal example, I was able to simply set this flag:

> LDFLAGS=-shared

In the code I was originally trying to compile (in an environment where LDFLAGS had values), I used:
> LDFLAGS="-shared $LDFLAGS"

I have hit this issue a few times (as have others). I have typically solved it by changing my environment (or I have added a main program). Would it be possible to detect if the `-shared` flag is needed for f2py? Should the gfortran_linux-64 library be adding this by default?

**Comment by pearu** [comment]
Sometimes one must have `-shared` (when building shared libraries) and sometimes one must not have it (when building static programs). Conda environment cannot foresee what users need. For instance, I have encountered a case when one must not have `-fPIC` in CFLAGS/CXXFLAGS but conda c/c++ compiler packages use `-fPIC` by default and it took me a while to figure out why the program occasionally crashes... 
In conclusion, if gfortran_linux-64 would use `-shared` by default, there will be complaints from people who do not need it. It is a problem with no good solution, IMHO.

If at all, the presence of `-shared` should be detected in numpy.distutils level but note that this flag is system/compiler-specific. Another approach would be to detect typical build failures and suggest possible fixes.

**Comment by pochedls** [comment]
Thanks for the explanation @pearu. I think this can be closed since it isn't a bug and it isn't clear how to deal with this issue more generally.

**Comment by durack1** [comment]
@pearu while I'm on this thread, what is the control that generates an output shared library named sharedLibrary.cpython-37m-x86_64-linux-gnu.so vs sharedLibrary.so?

It seems that with no apparent change to code, or the environment being used, the output filename tends to change.

**Comment by pearu** [comment]
@durack1 Python distutils would be my first guess.
